tests/core/templates/test_oa_hash_map.h

```diff
@@ -0,0 +1,193 @@
+	CHECK(lookup_res);
+	CHECK(value == 84);
+	CHECK(data == 84);
+}
+
+TEST_CASE("[OAHashMap] Set element") {
+	OAHashMap<int, int> map;
+	map.set(42, 84);
+	int data = 0;
+	bool lookup_res = map.lookup(42, data);
+	int value = *map.lookup_ptr(42);
+	CHECK(lookup_res);
+	CHECK(value == 84);
+	CHECK(data == 84);
+}
+
+TEST_CASE("[OAHashMap] Overwrite element") {
+	OAHashMap<int, int> map;
+	map.set(42, 84);
+	map.set(42, 1234);
+	int result = *map.lookup_ptr(42);
+	CHECK(result);
+}
+
+TEST_CASE("[OAHashMap] Remove element") {
+	OAHashMap<int, int> map;
+	map.insert(42, 84);
+	map.remove(42);
+	CHECK(!map.has(42));
+}
+
+TEST_CASE("[OAHashMap] Get Num_Elements") {
+	OAHashMap<int, int> map;
+	map.set(42, 84);
+	map.set(123, 84);
+	map.set(123, 84);
+	map.set(0, 84);
+	map.set(123485, 84);
+
+	CHECK(map.get_num_elements() == 4);
+}
+
+TEST_CASE("[OAHashMap] Iteration") {
+	OAHashMap<int, int> map;
+	map.insert(42, 84);
+	map.insert(123, 12385);
+	map.insert(0, 12934);
+	map.insert(123485, 1238888);
+	map.set(123, 111111);
+
+	Vector<Pair<int, int>> expected;
+	expected.push_back(Pair<int, int>(42, 84));
+	expected.push_back(Pair<int, int>(123, 111111));
+	expected.push_back(Pair<int, int>(0, 12934));
+	expected.push_back(Pair<int, int>(123485, 1238888));
+
+	int idx = 0;
+	for (OAHashMap<int, int>::Iterator it = map.iter(); it.valid; it = map.next_iter(it)) {
+		int64_t result = expected.find(Pair<int, int>(*it.key, *it.value));
+		CHECK(result >= 0);
+		++idx;
+	}
+}
+
+TEST_CASE("[OAHashMap] Insert, iterate, remove many strings") {
+	uint64_t pre_mem = Memory::get_mem_usage();
+	{
+		const int elem_max = 40;
+		OAHashMap<String, int> map;
+		for (int i = 0; i < elem_max; i++) {
+			map.insert(itos(i), i);
+		}
+
+		Vector<String> elems_still_valid;
+
+		for (int i = 0; i < elem_max; i++) {
+			if ((i % 5) == 0) {
+				map.remove(itos(i));
+			} else {
+				elems_still_valid.push_back(itos(i));
+			}
+		}
+
+		CHECK(elems_still_valid.size() == map.get_num_elements());
+
+		for (int i = 0; i < elems_still_valid.size(); i++) {
+			CHECK(map.has(elems_still_valid[i]));
+		}
+	}
+
+	CHECK(Memory::get_mem_usage() == pre_mem);
+}
+
+TEST_CASE("[OAHashMap] Clear") {
+	OAHashMap<int, int> map;
+	map.insert(42, 84);
+	map.insert(0, 1234);
+	map.clear();
+	CHECK(!map.has(42));
+	CHECK(!map.has(0));
+	CHECK(map.is_empty());
+}
+
+TEST_CASE("[OAHashMap] Non-trivial types") {
+	uint64_t pre_mem = Memory::get_mem_usage();
+	{
+		OAHashMap<String, Ref<Texture2D>> map1;
+		const uint32_t count = 10;
+		for (uint32_t i = 0; i < count; i++) {
+			String string = "qwerty";
+			string += i + '0';
+			Ref<Texture2D> ref_texture_2d;
+
+			map1.set(string, ref_texture_2d);
+			Ref<Texture2D> map_vec = *map1.lookup_ptr(string);
+			CHECK(map_vec == ref_texture_2d);
+		}
+		OAHashMap<String, Ref<Texture2D>> map1copy(map1);
+		CHECK(map1copy.has(String("qwerty") + '0'));
+		map1copy = map1;
+		CHECK(map1copy.has(String("qwerty") + 2 + '0'));
```

Comment:
What are you trying to achieve here? You're adding a character value, did you mean
```suggestion
		CHECK(map1copy.has(String("qwerty") + '2' + '0'));
```

Reply:
This should be  `CHECK(map1copy.has(String("qwerty") + '2'));`

Reply:
Or better `CHECK(map1copy.has(String("qwerty") + (2 + '0')));`

Reply:
Why though? It should just be `CHECK(map1copy.has(String("qwerty2"))`